README.md
# streamly-lz4

This library uses the LZ4 compression algorithm
<https://github.com/lz4/lz4> to compress and decompress a data stream
using [Haskell Streamly](https://streamly.composewell.com/).

## Running benchmarks

`download-corpora.sh` downloads and unpacks the [caterbury
corpus](https://corpus.canterbury.ac.nz/) used for benchmarking.

Run the following commands from the top level directory in the repo:
```
$ ./download-corpora.sh
$ cabal run bench-lz4 -- --quick
```

To use [fusion-plugin](https://github.com/composewell/fusion-plugin) for
benchmarks:

```
$ cabal run --flag fusion-plugin bench-lz4 -- --quick
```

### Benchmarking an external corpus

`BENCH_STREAMLY_LZ4_FILE` is looked up for file to benchmark.  This should
contain the absolute path of the target file.

`BENCH_STREAMLY_LZ4_STRATEGY` is looked up for which benchmarking combinator to
run on the target file.

It can contain values of the following structure,
- `c+speed+bufsize`
- `d+bufsize`
- `r+bufsize`

`c` corresponds to compress, `d` for decompress and `r` for resizing.

Example,
```
$ export BENCH_STREAMLY_LZ4_FILE="path/to/file/"
$ export BENCH_STREAMLY_LZ4_STRATEGY="c+400+640000"
$ cabal bench
```

The commands above runs the compression bench suite on **path/to/file/** with
the acceleration value of **500** read as arrays of size **640000** bytes.

*Note*: For the decompression and resizing bench suite a compressed file is
expected as input.


CHANGELOG.md
# Changelog

## 0.1.2 (Mar 2022)

* Support streamly-0.8.2

## 0.1.1 (Dec 2021)

* Support streamly-0.8.1

## 0.1.0

* Initial release


test/Main.hs
-- |
-- Module      : Main
-- Copyright   : (c) 2020 Composewell Technologies
-- License     : Apache-2.0
-- Maintainer  : streamly@composewell.com
-- Stability   : experimental
-- Portability : GHC
--
module Main (main) where

import Control.Monad (forM_)
import Control.Monad.IO.Class (MonadIO(..))
import Data.Word (Word8)
import Data.Function ((&))
import Streamly.Internal.Data.Stream.IsStream.Type (fromStreamD, toStreamD)
import System.IO (IOMode(..), openFile, hClose)
import System.IO.Temp (withSystemTempFile)
import Test.Hspec (describe, hspec, it, shouldBe)
import Test.QuickCheck (Property, forAll, property)
import Test.QuickCheck.Gen
    ( Gen, choose, elements, frequency, generate, listOf, vectorOf )
import Test.QuickCheck.Monadic (monadicIO)

import qualified Streamly.Internal.Data.Array.Foreign.Type as Array
import qualified Streamly.Internal.Data.Stream.IsStream as Stream
import qualified Streamly.Internal.FileSystem.Handle as Handle

import Streamly.Internal.LZ4.Config
import Streamly.Internal.LZ4
import Streamly.LZ4

-- Choose only 0 and 1 so that we can get good compressible sequences.
genArrayW8List :: Gen [Array.Array Word8]
genArrayW8List = listOf $ Array.fromList <$> listOf (elements [0,1])

genArrayW8Large :: Gen (Array.Array Word8)
genArrayW8Large = do
    let minArr = 1024 * 10
        maxArr = 1024 * 100
    arrS <- choose (minArr, maxArr)
    Array.fromList <$> vectorOf arrS (elements [0,1])

-- HC : Highly compressible
genArrayW8LargeHC :: (Int, Int) -> Gen (Array.Array Word8)
genArrayW8LargeHC (min_, max_) = do
    arrS <- choose (min_, max_)
    Array.fromList <$> vectorOf arrS (frequency [(9, pure 0), (1, pure 1)])

genNonEmptyListWith :: Gen (Array.Array Word8) -> Gen [Array.Array Word8]
genNonEmptyListWith gen = do
    sz <- choose (50, 100)
    vectorOf sz gen

genAcceleration :: Gen Int
genAcceleration = elements [-1..12]

decompressCompressChunk :: BlockConfig -> Int -> Array.Array Word8 -> IO ()
decompressCompressChunk conf i arr = do
    lz4Ctx <- c_createStream
    lz4CtxD <- c_createStreamDecode
    compressed <- compressChunk conf i lz4Ctx arr
    decompressed <- decompressChunk conf lz4CtxD compressed
    c_freeStream lz4Ctx
    c_freeStreamDecode lz4CtxD
    decompressed `shouldBe` arr

decompressCompressChunk2 ::
       BlockConfig -> Int -> Array.Array Word8 -> Array.Array Word8 -> IO ()
decompressCompressChunk2 conf i arr1 arr2 = do
    lz4Ctx <- c_createStream
    lz4CtxD <- c_createStreamDecode
    compressed1 <- compressChunk conf i lz4Ctx arr1
    compressed2 <- compressChunk conf i lz4Ctx arr2
    decompressed1 <- decompressChunk conf lz4CtxD compressed1
    decompressed2 <- decompressChunk conf lz4CtxD compressed2
    c_freeStream lz4Ctx
    c_freeStreamDecode lz4CtxD
    (decompressed1, decompressed2) `shouldBe` (arr1, arr2)

decompressResizedcompress :: BlockConfig -> Int -> [Array.Array Word8] -> IO ()
decompressResizedcompress conf i lst =
    let strm = Stream.fromList lst
     in do lst1 <-
               Stream.toList $ decompressChunksRaw $ compressChunks conf i strm
           lst `shouldBe` lst1

    where

    decompressChunksRaw = fromStreamD . decompressChunksRawD conf . toStreamD

decompressCompress :: BlockConfig -> Int -> Int -> [Array.Array Word8] -> IO ()
decompressCompress conf bufsize i lst = do
    let strm = Stream.fromList lst
    withSystemTempFile "LZ4" $ \tmp tmpH -> do
        compressChunks conf i strm & Handle.putChunks tmpH
        hClose tmpH
        lst1 <-
            Stream.toList
                $ Stream.bracket_ (openFile tmp ReadMode) hClose
                $ \h ->
                      Stream.unfold Handle.readChunksWithBufferOf (bufsize, h)
                          & decompressChunks conf
        lst1 `shouldBe` lst

{-# INLINE endMarkArr #-}
endMarkArr :: Array.Array Word8
endMarkArr = Array.fromListN 4 [0,0,0,0]

decompressCompressFrame ::
       BlockConfig -> FrameConfig -> Int -> Int -> [Array.Array Word8] -> IO ()
decompressCompressFrame bf ff bufsize i lst = do
    let strm = Stream.fromList lst
    withSystemTempFile "LZ4" $ \tmp tmpH -> do
        compressChunksFrame strm & Handle.putChunks tmpH
        hClose tmpH
        lst1 <-
            Stream.toList
                $ Stream.bracket_ (openFile tmp ReadMode) hClose
                $ \h ->
                      Stream.unfold Handle.readChunksWithBufferOf (bufsize, h)
                          & decompressChunksFrame
        lst1 `shouldBe` lst

    where

    decompressChunksFrame
        :: Stream.SerialT IO (Array.Array Word8)
        -> Stream.SerialT IO (Array.Array Word8)
    decompressChunksFrame =
        fromStreamD . decompressChunksRawD bf . resizeChunksD bf ff . toStreamD

    compressChunksFrame
        :: Stream.SerialT IO (Array.Array Word8)
        -> Stream.SerialT IO (Array.Array Word8)
    compressChunksFrame strm =
        if hasEndMark ff
        then (fromStreamD . compressChunksD bf i . toStreamD) strm
                 `Stream.append` Stream.fromPure endMarkArr
        else (fromStreamD . compressChunksD bf i . toStreamD) strm

decompressWithCompress :: Int -> Int -> [Array.Array Word8] -> IO ()
decompressWithCompress bufsize i lst = do
    -- Stream with header that returns the config
    -- Magic Little endian (4 bytes) = 407708164
    let magicLE = [4, 34, 77, 24]
    -- flg with everything unset
        flg = 64
    -- bd with 64KB block max size
        bd = 64
        headerChk = 0
        headerList = magicLE ++ [flg, bd, headerChk]
        header = Stream.fromList headerList
    headerArr <- Stream.fold (Array.writeN (length headerList)) header
    (bf, ff) <- Stream.parseD simpleFrameParserD header
    let strm = Stream.fromList lst
    withSystemTempFile "LZ4" $ \tmp tmpH -> do
        compressChunksFrame bf ff i strm
            & Stream.cons headerArr
            & Handle.putChunks tmpH
        hClose tmpH
        lst1 <-
            Stream.toList
                $ Stream.bracket_ (openFile tmp ReadMode) hClose
                $ \h ->
                      Stream.unfold Handle.readChunksWithBufferOf (bufsize, h)
                          & decompressWith_
        lst1 `shouldBe` lst

    where

    compressChunksFrame
        :: BlockConfig
        -> FrameConfig
        -> Int
        -> Stream.SerialT IO (Array.Array Word8)
        -> Stream.SerialT IO (Array.Array Word8)
    compressChunksFrame bf ff i_ strm =
        if hasEndMark ff
        then (fromStreamD . compressChunksD bf i_ . toStreamD) strm
                 `Stream.append` Stream.fromPure endMarkArr
        else (fromStreamD . compressChunksD bf i_ . toStreamD) strm

    decompressWith_ ::
           Stream.SerialT IO (Array.Array Word8)
        -> Stream.SerialT IO (Array.Array Word8)
    decompressWith_ =
        fromStreamD . decompressChunksWithD simpleFrameParserD . toStreamD

resizeIdempotence :: BlockConfig -> Property
resizeIdempotence conf =
    forAll ((,) <$> genAcceleration <*> genArrayW8List)
        $ \(acc, w8List) -> do
              let strm = compressChunks conf acc $ Stream.fromList w8List
              f1 <- Stream.toList $ resizeChunks strm
              f2 <- Stream.toList $ foldr ($) strm $ replicate acc resizeChunks
              f1 `shouldBe` f2

    where

    resizeChunks =
        fromStreamD . resizeChunksD conf defaultFrameConfig . toStreamD

main :: IO ()
main = do
    large <- generate $ genNonEmptyListWith genArrayW8Large
    largeHC <-
        generate
            $ genNonEmptyListWith
            $ genArrayW8LargeHC (1024 * 10, 1024 * 64)
    hspec $ do
        describe "Idempotence" $
            it "resize" (resizeIdempotence defaultBlockConfig)
        describe "Identity" $ do
            propsChunk defaultBlockConfig
            propsChunk2 defaultBlockConfig
            propsSimple defaultBlockConfig
            forM_ [-1, 5, 12, 100]
                $ \i ->
                      forM_ [1, 512, 32 * 1024, 256 * 1024]
                          $ \bufsize -> do
                                propsBig defaultBlockConfig bufsize i large
                                describe "Highly compressible"
                                    $ propsBig
                                          defaultBlockConfig bufsize i largeHC
            describe "setBlockMaxSize BlockMax256KB defaultBlockConfig" $ do
                let config = setBlockMaxSize BlockMax256KB defaultBlockConfig
                propsChunk config
                propsChunk2 config
                propsBig config 512 5 largeHC
            describe "defaultBlockConfig" $ do
                propsSimpleDecompressCompress defaultBlockConfig
                propsBigDecompressCompress defaultBlockConfig 512 5 large
                propsBigDecompressCompress defaultBlockConfig 512 5 largeHC
            describe "setFrameEndMark True defaultFrameConfig" $ do
                let ff = setFrameEndMark True defaultFrameConfig
                propsFrame defaultBlockConfig ff 512 5 large
                propsFrame defaultBlockConfig ff 512 5 largeHC
            describe "setFrameEndMark True . setBlockMaxSize BlockMax256KB"
                $ do
                    let bf = setBlockMaxSize BlockMax256KB defaultBlockConfig
                        ff = setFrameEndMark True defaultFrameConfig
                    propsFrame bf ff 512 5 large
                    propsFrame bf ff 512 5 largeHC
            describe "parse config (decompressWith)" $ do
                propsDecompressWithCompress 512 5 largeHC

    where

    propsSimpleDecompressCompress conf = do
        it "decompress . compress == id"
            $ property
            $ forAll
                  ((,) <$> genAcceleration <*> genArrayW8List)
                  (monadicIO . liftIO . uncurry (decompressCompress conf 512))

    propsBigDecompressCompress conf bufsize i l = do
        it
            ("decompress . compress ("
                 ++ show i ++ "/" ++ show bufsize ++ ") == id (big)")
            $ decompressCompress conf bufsize i l

    propsDecompressWithCompress bufsize i l = do
        it
            "decompressWith . compress == id"
            (decompressWithCompress bufsize i l)

    propsFrame bf ff bufsize i l = do
        it ("decompressFrame . compressFrame (" ++ show i ++ ") == id")
            $ decompressCompressFrame bf ff bufsize i l

    propsSimple conf = do
        it "decompressResized . compress == id"
            $ property
            $ forAll
                  ((,) <$> genAcceleration <*> genArrayW8List)
                  (monadicIO . liftIO . uncurry (decompressResizedcompress conf))
        it "decompress . compress == id"
            $ property
            $ forAll
                  ((,) <$> genAcceleration <*> genArrayW8List)
                  (monadicIO . liftIO . uncurry (decompressCompress conf 512))

    propsBig conf bufsize i l = do
        it ("decompressResized . compress (" ++ show i ++ ") == id (big)")
            $ decompressResizedcompress conf i l
        it
            ("decompress . compress ("
                 ++ show i ++ "/" ++ show bufsize ++ ") == id (big)")
            $ decompressCompress conf bufsize i l

    propsChunk2 conf = do
        it "decompressChunk . compressChunk (x2) == id (big)"
            $ property
            $ forAll
                  ((,,)
                       <$> genAcceleration
                       <*> genArrayW8Large
                       <*> genArrayW8Large)
                  (\(i, arr1, arr2) -> decompressCompressChunk2 conf i arr1 arr2)

    propsChunk conf = do
        it "decompressChunk . compressChunk == id (big)"
            $ property
            $ forAll
                  ((,) <$> genAcceleration <*> genArrayW8Large)
                  (uncurry (decompressCompressChunk conf))


decomp.hs
module Main (main) where

import Control.Monad (unless)
import Data.Semigroup (cycle1)
import Data.Word (Word8)
import Data.Function ((&))
import Streamly.Internal.Data.Array.Foreign (Array)
import Streamly.Internal.Data.Stream.StreamD (fromStreamD, toStreamD)
import Streamly.Prelude (SerialT)
import System.Directory (getCurrentDirectory, doesFileExist)
import System.Environment (lookupEnv)

import qualified Streamly.Internal.Data.Stream.IsStream as Stream
import qualified Streamly.Internal.Data.Stream.StreamD as StreamD
import qualified Streamly.Internal.FileSystem.File as File
import qualified Streamly.Internal.LZ4 as LZ4
import qualified Streamly.LZ4 as LZ4

import qualified Streamly.Internal.Data.Array.Foreign as Array
import qualified Streamly.Internal.Data.Producer.Source as Source
import qualified Streamly.Internal.Data.Producer as Producer
import qualified Streamly.Internal.Data.Producer.Type as Producer

{-
{-# INLINE decompressFrame #-}
decompressFrame :: String -> IO ()
decompressFrame corpus =
    let str = StreamD.unfold File.readChunksWithBufferOf (1024 * 64, corpus)
        -- prod = Producer.concat Producer.fromStreamD Array.producer
        -- srced = Source.producer prod
        srced = Source.producer Producer.fromStreamD
        unf = LZ4.decompressFrame srced
        -- seed = LZ4.ParsingBody $ Source.source $ Just $ Producer.OuterLoop str
        seed = LZ4.ParsingBody $ Source.source $ Just str
     in Stream.drain $ Stream.unfold (Producer.simplify unf) seed
     -}

{-# INLINE decompressFrame #-}
decompressFrame :: String -> IO ()
decompressFrame corpus =
    let str = StreamD.unfold File.readChunksWithBufferOf (1024 * 64, corpus)
        w8prod = Producer.concat_ Producer.fromStreamD Array.producer
        -- unf = LZ4.decompressFrame (Source.producer Producer.fromStreamD)
        pro = LZ4.decompressFrame (Source.producer w8prod)
     in Stream.drain $ Stream.produce pro (Source.source $ Just str)

-- nix-shell --run 'ghc -O2 -fdicts-strict -fspec-constr-recursive=16 -fmax-worker-args=16 -fplugin Fusion.Plugin -ddump-simpl -ddump-to-file -dsuppress-all Scratch.hs'

main :: IO ()
main = decompressFrame "corpora/large/bible.txt.normalized.compressed.frame"

{-
{-# INLINE _64KB #-}
_64KB :: Int
_64KB = 64 * 1024

main :: IO ()
main = do
    let corpus = "corpora/large/bible.txt.normalized.compressed.frame"
    let frameConf =
            LZ4.defaultConfig & LZ4.removeUncompressedSize (fromIntegral _64KB)
                & LZ4.addEndMark
    Stream.unfold File.readChunksWithBufferOf (1024 * 64, corpus)
        & (LZ4.decompress frameConf)
        & Stream.drain
-}


download-corpora.sh
#!/bin/sh

# This script downloads the corpora required to run certain tests and
# benchmarks.

BASE_DIR="corpora"
CORPORA="cantrbry artificl large"

# $1 = list of files to download from http://corpus.canterbury.ac.nz/resources/
#      without suffix
# $2 = directory to download and unzip (relative)
download_and_unzip() {
    # Make the base directory if it does not exist
    if test ! -d $2
    then
        echo "mkdir -p $2"
        mkdir -p $2
    fi

    for file in $1
    do
        # Remove existing files
        if test -d "$2/$file"
        then
            echo "rm -rf $2/$file"
            rm -rf "$2/$file"
        fi

        # Downloading tar
        if test ! -e "$2/${file}.tar.gz"
        then
            # Download
            echo "Downloading ${file}.tar.gz"
            wget -P $2 "http://corpus.canterbury.ac.nz/resources/${file}.tar.gz"
        else
            echo "${file}.tar.gz already exists, skipping"
        fi

        # Make directory
        echo "mkdir -p $2/$file"
        mkdir -p "$2/$file"

        # Extract files
        echo "Extracting $2/${file}.tar.gz"
        tar xf "$2/${file}.tar.gz" -C "$2/$file"
    done
}

download_and_unzip "$CORPORA" $BASE_DIR


src/Streamly/LZ4.hs
-- |
-- Module      : Streamly.LZ4
-- Copyright   : (c) 2020 Composewell Technologies
-- License     : Apache-2.0
-- Maintainer  : streamly@composewell.com
-- Stability   : experimental
-- Portability : GHC
--
-- Streaming APIs for LZ4 (<https://github.com/lz4/lz4>) compression and
-- decompression.
--
-- A compressed LZ4 object (e.g. a file) may be represented by a sequence of
-- one or more LZ4 frames defined by the [LZ4 frame
-- format](https://github.com/lz4/lz4/blob/dev/doc/lz4_Frame_format.md). A
-- frame consists of a frame header followed by a number of compressed blocks
-- and a frame footer. The frame header defines the attributes of the
-- compression method and the blocks in the frame. For example, the blocks may
-- be independently compressed or future blocks may depend on the past blocks.
-- It may also describe the maximum size of the blocks in the frame and use of
-- some optional features.
--
-- This module exposes combinators to only compress or decompress the stream of
-- blocks in a frame and not the frame itself. See the "Streamly.Internal.LZ4"
-- module for an experimental frame parsing function.
--
-- How the blocks are encoded, depends on the attributes specified in the frame
-- header.  We provide a 'BlockConfig' parameter to specify those options when
-- decoding or encoding a stream of blocks. Assuming you have parsed the frame,
-- you can set the 'BlockConfig' accordingly to parse the stream of blocks
-- appropriately.
--
-- Please build with
-- [fusion-plugin](https://hackage.haskell.org/package/fusion-plugin) for best
-- performance. See the [streamly build
-- guide](https://streamly.composewell.com/streamly-0.8.2/Compiling.html) for
-- more details.
--
-- The APIs are not yet stable and may change in future.
--
module Streamly.LZ4
    (
    -- * Configuration
      BlockConfig
    , defaultBlockConfig
    , BlockSize(..)
    , setBlockMaxSize

    -- * Combinators
    , compressChunks
    , decompressChunks
    )

where

import Control.Monad.IO.Class (MonadIO)
import Data.Word (Word8)
import Streamly.Internal.Data.Array.Foreign (Array)
import Streamly.Prelude (SerialT)
import Streamly.Internal.Data.Stream.IsStream.Type (fromStreamD, toStreamD)

import Streamly.Internal.LZ4.Config
import Streamly.Internal.LZ4

--------------------------------------------------------------------------------
-- Compression
--------------------------------------------------------------------------------

-- Note that "speedup" is specific to compression, therefore, it is not part of
-- BlockConfig.
--
-- XXX Ensure little-endian byte order as per the spec
-- XXX Check the size of the input array and fail if it is more than the max
-- block size set in BlockConfig.

-- | @compressChunks config speedup stream@ compresses an input stream of
-- @Array word8@ using the configuration defined by @config@. The resulting
-- stream is of type @Array Word8@ where each array represents a compressed
-- input block. Each input array becomes one compressed block.
--
-- @speedup@ is a compression speedup factor, more the value of @speedup@
-- faster the compression but the size of compressed data may increase. The
-- factor should be between 1 and 65537 inclusive, if it is less than 1 it is
-- set to 1 if it is more than 65537 then it is set to 65537.
--
-- LZ4 does not allow an uncompressed block of size more than 2,113,929,216
-- (0x7E000000) bytes (a little less than 2GiB).  If the compressed block
-- length is more than maximum uncompressed block length (approximately 2GiB)
-- it would result in a decompression error.
--
-- See 'BlockConfig' for more details about the format of the compressed block.
--
-- /Since 0.1.0/
{-# INLINE compressChunks #-}
compressChunks ::
       MonadIO m
    => BlockConfig
    -> Int
    -> SerialT m (Array Word8)
    -> SerialT m (Array Word8)
compressChunks cfg i m = fromStreamD (compressChunksD cfg i (toStreamD m))

--------------------------------------------------------------------------------
-- Decompression
--------------------------------------------------------------------------------

-- | Decompress a stream of @Array Word8@ compressed using LZ4 stream
-- compression. See 'compressChunks' for the format of the input blocks. The
-- input chunks could be of any size, they are resized to the appropriate block
-- size before decompression based on block headers. The decompressed output
-- arrays correspond to one compressed block each.
--
-- /Since 0.1.0/
{-# INLINE decompressChunks #-}
decompressChunks ::
       MonadIO m
    => BlockConfig
    -> SerialT m (Array Word8)
    -> SerialT m (Array Word8)
decompressChunks bf =
    fromStreamD
        . decompressChunksRawD bf
        . resizeChunksD bf defaultFrameConfig . toStreamD


benchmark/Main.hs
-- |
-- Module      : Main
-- Copyright   : (c) 2020 Composewell Technologies
-- License     : Apache-2.0
-- Maintainer  : streamly@composewell.com
-- Stability   : experimental
-- Portability : GHC
--
module Main (main) where

import Control.Monad (unless)
import Data.Semigroup (cycle1)
import Data.Word (Word8)
import Data.Function ((&))
import Streamly.Internal.Data.Array.Foreign (Array)
import Streamly.Internal.Data.Stream.IsStream.Type (fromStreamD, toStreamD)
import Streamly.Prelude (SerialT)
import System.Directory (getCurrentDirectory, doesFileExist)
import System.Environment (lookupEnv)

import qualified Streamly.Internal.Data.Stream.IsStream as Stream
import qualified Streamly.Internal.Data.Array.Foreign as Array
import qualified Streamly.Internal.FileSystem.File as File
import qualified Streamly.Internal.LZ4 as LZ4
import qualified Streamly.Internal.LZ4.Config as LZ4
import qualified Streamly.LZ4 as LZ4

import Gauge.Main

{-# INLINE _64KB #-}
_64KB :: Int
_64KB = 64 * 1024

{-# INLINE _10MB #-}
_10MB :: Int
_10MB = 10 * 1024 * 1024

--------------------------------------------------------------------------------
-- Corpora helpers
--------------------------------------------------------------------------------

{-# INLINE large_bible_txt #-}
large_bible_txt :: String
large_bible_txt = "large/bible.txt"

{-# INLINE large_world192_txt #-}
large_world192_txt :: String
large_world192_txt = "large/world192.txt"

{-# INLINE cantrbry_alice29_txt #-}
cantrbry_alice29_txt :: String
cantrbry_alice29_txt = "cantrbry/alice29.txt"

--------------------------------------------------------------------------------
-- Bootstrapping
--------------------------------------------------------------------------------

normalizedName :: String -> String
normalizedName x = x ++ ".normalized"

bigCompressedName :: String -> String
bigCompressedName x = x ++ ".normalized.compressed.big"

smallCompressedName :: String -> String
smallCompressedName x = x ++ ".normalized.compressed.small"

withCompressedName :: String -> String
withCompressedName x = x ++ ".normalized.compressed.frame"

bootstrap :: String -> IO ()
bootstrap fp = do
    let normalizedFp = normalizedName fp
        compressedFpBig = bigCompressedName fp
        compressedFpSmall = smallCompressedName fp
        compressedWith = withCompressedName fp
    fileExists <- doesFileExist normalizedFp
    unless fileExists $ do
        putStrLn $ "Normalizing " ++ fp
        let fileStream = Stream.unfold File.read fp
            combinedStream =
                Stream.arraysOf _64KB
                    $ Stream.take _10MB
                    $ cycle1 fileStream
        combinedStream & File.fromChunks normalizedFp
        combinedStream & LZ4.compressChunks LZ4.defaultBlockConfig 65537
                       & File.fromChunks compressedFpBig
        combinedStream & LZ4.compressChunks LZ4.defaultBlockConfig 1
                       & File.fromChunks compressedFpSmall
        -- Stream with header that returns the config
        -- Magic Little endian (4 bytes) = 407708164
        let magicLE = [4, 34, 77, 24]
        -- flg with everything unset
            flg = 64
        -- bd with 64KB block max size
            bd = 64
            headerChk = 0
            headerList = magicLE ++ [flg, bd, headerChk]
            header = Stream.fromList headerList
        headerArr <- Stream.fold (Array.writeN (length headerList)) header
        (bf, ff) <- Stream.parseD LZ4.simpleFrameParserD header
        combinedStream & compressChunksFrame bf ff 65537
                       & Stream.cons headerArr
                       & File.fromChunks compressedWith

    where

    {-# INLINE endMarkArr #-}
    endMarkArr :: Array.Array Word8
    endMarkArr = Array.fromListN 4 [0,0,0,0]

    compressChunksFrame
        :: LZ4.BlockConfig
        -> LZ4.FrameConfig
        -> Int
        -> Stream.SerialT IO (Array.Array Word8)
        -> Stream.SerialT IO (Array.Array Word8)
    compressChunksFrame bf ff i_ strm =
        if LZ4.hasEndMark ff
        then (fromStreamD . LZ4.compressChunksD bf i_ . toStreamD) strm
                 `Stream.append` Stream.fromPure endMarkArr
        else (fromStreamD . LZ4.compressChunksD bf i_ . toStreamD) strm


--------------------------------------------------------------------------------
-- Benchmark helpers
--------------------------------------------------------------------------------

type Combinator = SerialT IO (Array Word8) -> SerialT IO (Array Word8)

{-# INLINE benchCorpus #-}
benchCorpus :: Int -> String -> String -> Combinator -> Benchmark
benchCorpus bufsize name corpus combinator =
    let bname = ("bufsize(" ++ show bufsize ++ ")/" ++ name ++ "/" ++ corpus)
     in bench bname $ nfIO $ do
            Stream.unfold File.readChunksWithBufferOf (bufsize, corpus)
                & combinator
                & Stream.drain

--------------------------------------------------------------------------------
-- Benchmarks
--------------------------------------------------------------------------------

{-# INLINE compress #-}
compress :: Int -> Int -> String -> Benchmark
compress bufsize i corpus =
    benchCorpus bufsize ("compress " ++ show i) corpus
        $ LZ4.compressChunks LZ4.defaultBlockConfig i

{-# INLINE decompress #-}
decompress :: Int -> String -> Benchmark
decompress bufsize corpus =
    benchCorpus bufsize "decompress" corpus
        $ LZ4.decompressChunks LZ4.defaultBlockConfig

{-# INLINE decompressWith #-}
decompressWith :: Int -> String -> Benchmark
decompressWith bufsize corpus =
    benchCorpus bufsize "decompressWith" corpus
        $ fromStreamD
        . LZ4.decompressChunksWithD LZ4.simpleFrameParserD . toStreamD

{-# INLINE resize #-}
resize :: Int -> String -> Benchmark
resize bufsize corpus =
    benchCorpus bufsize "resize" corpus
        $ fromStreamD
        . LZ4.resizeChunksD LZ4.defaultBlockConfig LZ4.defaultFrameConfig
        . toStreamD

--------------------------------------------------------------------------------
-- Reading environment
--------------------------------------------------------------------------------

-- Environment variables looked up
-- BENCH_STREAMLY_LZ4_FILE
-- BENCH_STREAMLY_LZ4_STRATEGY
-- - c+speed+buffer
-- - d+buffer
-- - r+buffer
--
-- Example:
-- > export BENCH_STREAMLY_LZ4_FILE="path/to/file/"
-- > export BENCH_STREAMLY_LZ4_STRATEGY="c+400+640000"
-- > cabal bench
--
-- The above commands will benchmark file on compression with acceleration value
-- of 400 and buffer size of 640000

data Strategy
    = Compress Int Int
    | Decompress Int
    | Resize Int

{-# INLINE parseStrategy #-}
parseStrategy :: String -> Strategy
parseStrategy ('c':_:r) =
    let (speed, pbufsize) = span (/= '+') r
     in Compress (read speed) (read (tail pbufsize))
parseStrategy ('d':_:r) = Decompress (read r)
parseStrategy ('r':_:r) = Resize (read r)
parseStrategy _ = error "Cannot parse BENCH_STREAMLY_LZ4_STRATEGY"

{-# INLINE runStrategy #-}
runStrategy :: String -> Strategy -> Benchmark
runStrategy file (Compress speed bufsize) = compress bufsize speed file
runStrategy file (Decompress bufsize) = decompress bufsize file
runStrategy file (Resize bufsize) = resize bufsize file

{-# INLINE tryBenchExternal #-}
tryBenchExternal :: IO (Maybe Benchmark)
tryBenchExternal = do
    fr <- lookupEnv "BENCH_STREAMLY_LZ4_FILE"
    case fr of
        Nothing -> return Nothing
        Just file -> do
            fs <- lookupEnv "BENCH_STREAMLY_LZ4_STRATEGY"
            return
                $ case fs of
                      Nothing -> Nothing
                      Just s -> Just $ runStrategy file (parseStrategy s)

--------------------------------------------------------------------------------
-- Main
--------------------------------------------------------------------------------

main :: IO ()
main = do
    base <- flip (++) "/corpora/" <$> getCurrentDirectory
    let rel f = base ++ f
        relNormalized f = normalizedName (base ++ f)
        relBigCompressed f = bigCompressedName (base ++ f)
        relWithCompressed f = withCompressedName (base ++ f)
        relSmallCompressed f = smallCompressedName (base ++ f)
    bootstrap (rel large_bible_txt)
    bootstrap (rel large_world192_txt)
    bootstrap (rel cantrbry_alice29_txt)
    external <- maybe [] (: []) <$> tryBenchExternal
    defaultMain $
        external ++
        [ compression_files relNormalized
        , decompression_files_big relBigCompressed
        , decompression_files_with relWithCompressed
        , decompression_files_small relSmallCompressed
        , compression_accelaration relNormalized
        , compression_buffer relNormalized
        , decompression_buffer relBigCompressed
        , resizing_buffer relBigCompressed
        ]

    where

    compression_files f =
        bgroup
            "compress/files"
            [ compress _64KB 5 (f large_bible_txt)
            , compress _64KB 5 (f large_world192_txt)
            , compress _64KB 5 (f cantrbry_alice29_txt)
            ]

    decompression_files_big f =
        bgroup
            "decompress/files/big"
            [ decompress _64KB (f large_bible_txt)
            , decompress _64KB (f large_world192_txt)
            , decompress _64KB (f cantrbry_alice29_txt)
            ]

    decompression_files_with f =
        bgroup
            "decompressWith"
            [ decompressWith _64KB (f large_bible_txt)
            , decompressWith _64KB (f large_world192_txt)
            , decompressWith _64KB (f cantrbry_alice29_txt)
            ]

    decompression_files_small f =
        bgroup
            "decompression/files/small"
            [ decompress _64KB (f large_bible_txt)
            , decompress _64KB (f large_world192_txt)
            , decompress _64KB (f cantrbry_alice29_txt)
            ]

    compression_accelaration f =
        bgroup
            "compression/acceleration"
            [ compress _64KB (-1) (f large_bible_txt)
            , compress _64KB 10 (f large_bible_txt)
            , compress _64KB 1000  (f large_bible_txt)
            , compress _64KB 65537 (f large_bible_txt)
            ]

    compression_buffer f =
        bgroup
            "compression/buffer"
            [ compress (_64KB `div` 10) 5 (f large_bible_txt)
            , compress _64KB 5 (f large_bible_txt)
            , compress (_64KB * 10) 5 (f large_bible_txt)
            ]

    decompression_buffer f =
        bgroup
            "decompression/buffer"
            [ decompress (_64KB `div` 10) (f large_bible_txt)
            , decompress _64KB (f large_bible_txt)
            , decompress (_64KB * 10) (f large_bible_txt)
            ]

    resizing_buffer f =
        bgroup
            "resizing/buffer"
            [ resize (_64KB `div` 10) (f large_bible_txt)
            , resize _64KB (f large_bible_txt)
            , resize (_64KB * 10) (f large_bible_txt)
            ]
